---
title: Multi-step correction of batch effects in single-cell RNA-seq data
author: 
- name: Aaron T. L. Lun
  affiliation: Cancer Research UK Cambridge Institute, Li Ka Shing Centre, Robinson Way, Cambridge CB2 0RE, United Kingdom
- name: Michael D. Morgan
  affiliation: Wellcome Trust Sanger Institute, Wellcome Genome Campus, Hinxton, Cambridge CB10 1SA, United Kingdom
date: "`r Sys.Date()`"
vignette: >
  %\VignetteIndexEntry{11. Advanced batch correction}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}    
output: 
  BiocStyle::html_document:
    titlecaps: false
    toc_float: true
bibliography: ref.bib
---

```{r style, echo=FALSE, results='hide', message=FALSE, cache=FALSE}
library(BiocStyle)
library(knitr)
opts_chunk$set(error=FALSE, message=FALSE, warning=FALSE, cache=TRUE)
opts_chunk$set(fig.asp=1)
```

```{r, cache=FALSE, echo=FALSE, results="hide"}
simpleSingleCell:::.compile("batch") # pancreas
```

# Introduction

In the `r Biocpkg("simpleSingleCell", "batch.html", "previous batch correction workflow")`, we merged two scRNA-seq data sets involving human pancreas cell populations.
This workflow extends the previous example to describe how to perform a more complicated merge operation involving different levels of batch effects.
Here, we will use pancreas data sets generated using Smart-based technologies [@segerstolpe2016singlecell;@lawlor2017singlecell],
and merge them with the previous data sets generated using CEL-seq-based methods [@grun2016denovo;@muraro2016singlecell].
Our overall strategy is to use a hierarchical merge to remove batch effects within each technology,
followed by the removal of batch effects between technologies.

# Loading in the data

## SMARTer, GSE86469

### Reading in data

Here, we use data from the @lawlor2017singlecell study of pancreatic islet cells from healthy and type II diabetic donors.
This was generated using the SMARTer protocol on the Fluidigm C1 system.
We download and cache the count matrix using the `r Biocpkg("BiocFileCache")` package.

```{r}
library(BiocFileCache)
bfc <- BiocFileCache(ask=FALSE)    
count.tab <- bfcrpath(bfc, file.path(
    "ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE86nnn/GSE86469/suppl",
    "GSE86469_GEO.islet.single.cell.processed.data.RSEM.raw.expected.counts.csv.gz"
))
```

We read in the count matrix in sparse format using `r Biocpkg("scater")` [@mccarthy2017scater].

```{r}
library(scater)
mat <- readSparseCounts(count.tab, sep=",", quote='"', row.names=1)
dim(mat)
```

We load in the metadata from NCBI GEO using the `r Biocpkg("GEOquery")` package [@davis2007geoquery].

```{r}
library(GEOquery)
metadata <- pData(getGEO("GSE86469")[[1]])
metadata <- metadata[,c("title", "cell type:ch1", "islet unos id:ch1")]

rownames(metadata) <- metadata$title
metadata <- metadata[,-1]
colnames(metadata) <- c("CellType", "Donor")

stopifnot(identical(colnames(mat), rownames(metadata)))
head(metadata)
```

Finally, we create a `SingleCellExperiment` object.

```{r}
library(SingleCellExperiment)
sce.gse86469 <- SingleCellExperiment(list(counts=mat), colData=metadata)
isSpike(sce.gse86469, "ERCC") <- grep("^ERCC-", rownames(sce.gse86469))
sce.gse86469
```

### Quality control and normalization

We remove low-quality cells based on outliers for various quality control metrics,
such as the total library size and the number of expressed genes.
This is similar to what was described `r Biocpkg("simpleSingleCell", "reads.html#3_quality_control_on_the_cells", "previously")`.
Note that this data does not contain any counts for spike-in transcripts, 
so the spike-in percentage is not used here.

```{r}
sce.gse86469 <- calculateQCMetrics(sce.gse86469, compact=TRUE)
qc.mat <- cbind(
    NFeatures=isOutlier(sce.gse86469$scater_qc$all$total_features_by_counts, 
        log=TRUE, type="lower", nmads=3),
    LibSize=isOutlier(sce.gse86469$scater_qc$all$total_counts, 
        log=TRUE, type="lower", nmads=3)
)
colSums(qc.mat)
discard <- rowMeans(qc.mat) > 0
sce.gse86469 <- sce.gse86469[,!discard]
summary(discard)
```

We compute size factors with the deconvolution method from the `r Biocpkg("scran")` package [@lun2016pooling].
Pre-clustering is performed using `quickCluster()` to avoid pooling together very different cells.
Note the use of `IrlbaParam()` from `r Biocpkg("BiocSingular")` to speed up the PCA calculations.

```{r gse}
library(scran)
library(BiocSingular)
clusters <- quickCluster(sce.gse86469, BSPARAM=IrlbaParam())
table(clusters)
sce.gse86469 <- computeSumFactors(sce.gse86469, clusters=clusters)
summary(sizeFactors(sce.gse86469))
```

There is no need for spike-in normalization, as there are actually no spike-in counts. 
We thus proceed directly to calculation of the log-transformed normalized expression values for endogenous genes.

```{r}
# Ignore warnings due to no spike-in size factors.
suppressWarnings(sce.gse86469 <- normalize(sce.gse86469))
```

### Modelling variability
    
Given that no spike-ins are available, we fit a mean-dependent trend to the variances of the endogenous genes to model technical noise.
This requires the assumptions that have been stated `r Biocpkg("simpleSingleCell", "var.html#when-spike-ins-are-unavailable", "elsewhere")`.

```{r}
fit <- trendVar(sce.gse86469, use.spikes=FALSE, 
    block=sce.gse86469$Donor, loess.args=list(span=0.05))
dec.gse86469 <- decomposeVar(sce.gse86469, fit)
dec.gse86469[order(dec.gse86469$bio, decreasing=TRUE),]
```

Figure \@ref(fig:var-gse86469) shows the strong mean-variance relationship that is typical of read count data.

```{r var-gse86469, fig.cap="Variance of normalized log-expression values for each gene in the GSE86469 data set, plotted against the mean log-expression. The blue line represents the mean-dependent trend fitted to the variances of all genes."}
plot(fit$mean, fit$var, xlab="Mean log-expression",
    ylab="Variance of log-expression")
curve(fit$trend(x), col="dodgerblue", add=TRUE, lwd=2)
```

## Smart-seq2, E-MTAB-5061

### Reading in the data

Here, we use data from the @@segerstolpe2016singlecell study.
The good news is that the authors have provided a count table in the ArrayExpress entry for this project.
We download it using `r Biocpkg("BiocFileCache")` to cache the results:

```{r}
bfc <- BiocFileCache(ask=FALSE)    
emat <- bfcrpath(bfc, file.path("https://www.ebi.ac.uk/arrayexpress",
    "experiments/E-MTAB-5061/files/E-MTAB-5061.processed.1.zip"))
count.file <- "pancreas_refseq_rpkms_counts_3514sc.txt"
```

The bad news is that the count table is needlessly complex:

- The first 2 columns contain the gene symbol and NCBI GenBank transcript identifiers for each row.
- The next `X` columns are the RPKMs, for `X` cells.
- The remaining `X` columns are the counts.

This requires some additional work to extract the useful data.
The first line contains the names of the cells, 
so we can use this to determine the number and indices of the columns with per-cell counts.

```{r}
col.names <- read.table(unz(emat, count.file), header=FALSE, sep="\t", 
    stringsAsFactors=FALSE, comment.char="", nrows = 1)[,-1]
ncells <- length(col.names)

what <- vector("list", ncells*2 + 2)
what[[1]] <- "character"
what[seq_len(ncells) + ncells + 2] <- "integer"
```

We then read in the gene symbols and the counts.
We use the gene symbols as the GenBank IDs have been rather clumsily formatted.

```{r}
emtab.df <- read.table(unz(emat, count.file), header=FALSE, sep="\t", 
    stringsAsFactors=FALSE, colClasses=what, skip=1)
row.names <- emtab.df[,1]
emtab.df <- emtab.df[,-1]
colnames(emtab.df) <- col.names
dim(emtab.df)
```

Some work is required to translate the symbols into Ensembl gene identifiers via the `r Biocpkg("org.Hs.eg.db")` package.

```{r}
library(org.Hs.eg.db)
ens.id <- mapIds(org.Hs.eg.db, keys=row.names, 
    keytype="SYMBOL", column="ENSEMBL")
ens.id <- ifelse(is.na(ens.id), row.names, ens.id)

keep <- !duplicated(ens.id)
emtab.df <- emtab.df[keep,]
rownames(emtab.df) <- ens.id[keep]
head(rownames(emtab.df))
```

We read in the metadata and extract the appropriate columns.

```{r}
meta.fname <- bfcrpath(bfc, file.path("https://www.ebi.ac.uk/arrayexpress",
    "files/E-MTAB-5061/E-MTAB-5061.sdrf.txt"))
emtab.sdrf <- read.delim(meta.fname, stringsAsFactors=FALSE)
stopifnot(identical(sort(emtab.sdrf$Source.Name), sort(colnames(emtab.df))))    

emtab.sdrf <- emtab.sdrf[match(colnames(emtab.df), emtab.sdrf$Source.Name),]
emtab.meta <- emtab.sdrf[, c("Assay.Name", 
    "Characteristics.cell.type.", "Characteristics.individual.",
    "Characteristics.single.cell.well.quality.")]
colnames(emtab.meta) <- c("Sample", "CellType", "Donor", "Quality")
emtab.meta$Study <- "E-MTAB-5061"
head(emtab.meta)
```

Some editing of the cell type labels is necessary for consistency with GSE86469.

```{r}
emtab.meta$CellType <- gsub(" cell", "", emtab.meta$CellType)
emtab.meta$CellType <- paste0(
    toupper(substr(emtab.meta$CellType, 1, 1)),
    substring(emtab.meta$CellType, 2))
table(emtab.meta$CellType)
```

Finally, we create a `SingleCellExperiment` object.

```{r}
sce.emtab <- SingleCellExperiment(list(counts=as.matrix(emtab.df)), 
    colData=emtab.meta)
isSpike(sce.emtab, "ERCC") <- grep("^ERCC_", rownames(sce.emtab))
sce.emtab
```

### Quality control and normalization

We first remove the low quality cells that were marked by the authors.

```{r}
low.qual <- sce.emtab$Quality == "low quality cell"
sce.emtab <- sce.emtab[,!low.qual]
summary(low.qual)
```

We also remove low quality cells based on our own quality control metrics.
It is debatable whether these two separate rounds of quality control are necessary, 
but we do this for consistency with respect to the preprocessing performed across all data sets.

```{r}
sce.emtab <- calculateQCMetrics(sce.emtab, compact=TRUE)
qc.mat <- cbind(
    NFeatures=isOutlier(sce.emtab$scater_qc$all$total_features_by_counts, 
        log=TRUE, type="lower", nmads=3),
    LibSize=isOutlier(sce.emtab$scater_qc$all$total_counts, 
        log=TRUE, type="lower", nmads=3),
    SpikePct=isOutlier(sce.emtab$scater_qc$feature_control_ERCC$pct_counts, 
        type="higher", nmads=3)
)
colSums(qc.mat)
discard <- rowMeans(qc.mat) > 0
sce.emtab <- sce.emtab[,!discard]
summary(discard)
```

We compute size factors using the pre-clustering and deconvolution approach.

```{r}
clusters <- quickCluster(sce.emtab, BSPARAM=IrlbaParam())
table(clusters)
sce.emtab <- computeSumFactors(sce.emtab, clusters=clusters)
summary(sizeFactors(sce.emtab))
```

We also compute separate size factors for the spike-in counts.
Note that some cells have no spike-in counts and will not be useful for downstream steps that rely on spike-ins.

```{r}
sce.emtab <- computeSpikeFactors(sce.emtab, general.use=FALSE)
summary(sizeFactors(sce.emtab, "ERCC"))
```

### Modelling variability

Variance modelling requires some care in this data set.
The mean-variance trend is highly variable across cell populations from different donors,
presumably because each donor was processed in a separate batch.
Thus, we have to block on `Donor` using the `multiBlockVar()` function as discussed `r Biocpkg("simpleSingleCell", "var.html#fitting-batch-specific-trends", "here")`.
We also have to remove cells with no spike-ins, as they are not useful for modelling technical noise;
and remove cells from donor `AZ`, which has very low spike-in concentrations.

```{r}
for.hvg <- sce.emtab[,sizeFactors(sce.emtab, "ERCC") > 0
    & sce.emtab$Donor!="AZ"]
for.hvg <- multiBlockNorm(for.hvg, for.hvg$Donor) 
dec.emtab <- multiBlockVar(for.hvg, for.hvg$Donor)
head(dec.emtab[order(dec.emtab$bio,decreasing=TRUE),-7])
```

Figure \@ref(fig:var-emtab) demonstrates the diversity of mean-variance relationships across different donors.

```{r var-emtab, fig.wide=TRUE, fig.asp=1, fig.cap="Variance of normalized log-expression values for each gene in each donor of the EMTAB-5061 data set, plotted against the mean log-expression. The blue line represents the mean-dependent trend fitted to the variances of spike-in transcripts within each donor."}
all.donors <- unique(for.hvg$Donor)
par(mfrow=c(ceiling(length(all.donors)/3),3))
is.spike <- isSpike(for.hvg)
for (plate in all.donors) {
    cur.out <- dec.emtab$per.block[[plate]]
    plot(cur.out$mean, cur.out$total, pch=16, cex=0.6, main=plate, 
        xlab="Mean log-expression", ylab="Variance of log-expression")
    curve(metadata(cur.out)$trend(x), col="dodgerblue", lwd=2, add=TRUE)
    points(cur.out$mean[is.spike], cur.out$total[is.spike], col="red", pch=16)
}
```

# Feature selection across batches

Recall that our aim is to merge data from all four pancreas data sets [@segerstolpe2016singlecell;@lawlor2017singlecell;grun2016denovo;@muraro2016singlecell].
To do so, we first load in the CEL-seq data sets that we processed `r Biocpkg("simpleSingleCell", "batch.html", "previously")`.

```{r}
sce.gse81076 <- readRDS("gse81076_sce.rds")
dec.gse81076 <- readRDS("gse81076_dec.rds") 
sce.gse85241 <- readRDS("gse85241_sce.rds") 
dec.gse85241 <- readRDS("gse85241_dec.rds") 
```

We define the universe of genes that are common across all batches.
This is made straightforward by the presence of common Ensembl identifiers.

```{r}
universe <- Reduce(intersect, list(rownames(dec.gse81076), 
    rownames(dec.gse85241), rownames(dec.gse86469), 
    rownames(dec.emtab)))
universe <- universe[!grepl("^ERCC-", universe)] # removing spike-ins.
length(universe)
```

We adjust the size factors with `multiBatchNorm()` to make them more comparable across batches.
This mitigates differences in scale and variance in the log-expression values between batches, especially between technologies.

```{r}
library(batchelor)
nout <- multiBatchNorm(sce.gse81076[universe,], sce.gse85241[universe,],
    sce.gse86469[universe,], sce.emtab[universe,])
sce.gse81076 <- nout[[1]]
sce.gse85241 <- nout[[2]]
sce.gse86469 <- nout[[3]]
sce.emtab <- nout[[4]]
```

We keep all genes with positive average biological components across all batches.
This is a relaxed approach to feature selection that ensures that interesting features in one or more batches are retained.

```{r}
mean.bio <- (dec.gse81076[universe,"bio"] + dec.gse85241[universe,"bio"]
    + dec.gse86469[universe,"bio"] + dec.emtab[universe,"bio"])/4
chosen <- universe[mean.bio > 0]
length(chosen)
```

We then subset all of the `SingleCellExperiment` objects so that only these features of interest are retained.

```{r}
sce.gse81076 <- sce.gse81076[chosen,]
sce.gse85241 <- sce.gse85241[chosen,]
sce.gse86469 <- sce.gse86469[chosen,]
sce.emtab <- sce.emtab[chosen,]
```

# Multi-batch principal components analysis

We use the `multiBatchPCA()` function to perform a PCA across _all_ batches to be merged.
This ensures that all cells are placed onto the same coordinate space, which would obviously not be possible if a PCA was performed for each batch separately.
Specifically, `multiBatchPCA()` performs a modified PCA to ensure that each supplied matrix contributes equally to the definition of the PC space.
This avoids problems with imbalances in the number of cells across batches, meaning that smaller batches (possibly with unique cell types) are not ignored.

```{r}
set.seed(1000)
pcs <- multiBatchPCA(
    gse81076=sce.gse81076,
    gse85241=sce.gse85241,
    gse86469=sce.gse86469,
    emtab=sce.emtab,
    BSPARAM=IrlbaParam(deferred=TRUE)
)
names(pcs)
```

Typical applications of `fastMNN()` will automatically call the `multiBatchPCA()` function on gene expression inputs.
However, this is not appropriate here as we will be performing a hierarchical merge.
Each call to `fastMNN()` will only involve a subset of batches,
and it would be difficult to try to merge results from two separate PCAs involving different subsets of data.
We need to run `multiBatchPCA()` manually on all batches to ensure that they are on the same coordinate system during merging.

**Comments from Aaron:**

- The `IrlbaParam(deferred=TRUE)` setting instructs `multiBatchPCA()` to perform a fast approximate PCA with methods from the `r CRANpkg("irlba")` package.
This involves some randomization and thus requires the seed to be set to obtain reproducible results.
- Here, we have applied `multiBatchPCA()` to the batch-level inputs for convenience.
It is also possible to supply donor-level matrices to equalize contributions across donors,
but this requires a bit more data manipulation that we will omit for the sake of simplicity.
- For full consistency with the `fastMNN()` defaults, we would call `cosineNorm()` on each log-expression matrix prior to running `multiBatchPCA()`. 
However, this is not technically necessary as all batches should be on the same scale already (see `?cosineNorm` for a discussion of this).

# Hierarchical merging

## Merging the Smart-based data

The @segerstolpe2016singlecell study contains strong donor effects that interfere with correction.
Several cell types exhibit strong per-donor effects such that the multiple clusters cannot be fully merged with the corresponding single cluster in GSE86469 (Figure \@ref(fig:smart-raw)).

```{r smart-raw, fig.wide=TRUE, fig.asp=0.5, fig.cap="t-SNE plots of the merged Smart-based data sets. Each point is a cell coloured by the batch of origin (left) or the donor of origin in E-MTAB-5061 (right)."}
direct.smart <- fastMNN(pcs$gse86469, pcs$emtab, k=20, pc.input=TRUE)

set.seed(2000)
tsne.out <- Rtsne::Rtsne(direct.smart$corrected, perplexity=30, pca=FALSE)
df <- data.frame(x=tsne.out$Y[,1], y=tsne.out$Y[,2], 
    batch=rep(c("GSE86469", "E-MTAB-5061"), 
        c(ncol(sce.gse86469), ncol(sce.emtab))),
    donor=c(rep("unknown", ncol(sce.gse86469)), sce.emtab$Donor))

multiplot(
    ggplot(df) + geom_point(aes(x=x, y=y, color=donor)) +
        xlab("t-SNE 1") + ylab("t-SNE 2") + ggtitle("By donor"),
    ggplot(df) + geom_point(aes(x=x, y=y, color=batch)) +
        xlab("t-SNE 1") + ylab("t-SNE 2") + ggtitle("By batch"),
    cols=2)
```

To overcome this, the first step of our hierarchical merge is to remove differences between donors _within_ E-MTAB-5061.
This simply involves calling `fastMNN()` with the `batch=` argument set to the `Donor` variable. 
Note the use of `pc.input=TRUE` to avoid performing a PCA on a matrix of low-dimensional coordinates.

```{r}
fixed.emtab <- fastMNN(pcs$emtab, batch=sce.emtab$Donor, k=20, pc.input=TRUE)
```

It is then straightforward to merge these corrected expression values with the data from GSE86469.
This removes differences between studies and represents the second level of the merge hierarchy.

```{r}
mnn.smart <- fastMNN(pcs$gse86469, fixed.emtab, k=20, pc.input=TRUE)
```

This strategy eliminates the donor-based structure in the merged data (Figure \@ref(fig:smart-fixed)),
collapsing the previously distinct per-donor clusters into a single entity.

```{r smart-fixed, fig.wide=TRUE, fig.asp=0.5, fig.cap="t-SNE plots of the merged Smart-based data sets. Each point is a cell coloured by the batch of origin (left) or the donor of origin in E-MTAB-5061 (right)."}
set.seed(2000)
tsne.out <- Rtsne::Rtsne(mnn.smart$corrected, perplexity=30, pca=FALSE)
df <- data.frame(x=tsne.out$Y[,1], y=tsne.out$Y[,2], 
    batch=rep(c("GSE86469", "E-MTAB-5061"), 
        c(ncol(sce.gse86469), ncol(sce.emtab))),
    donor=c(rep(NA, ncol(sce.gse86469)), sce.emtab$Donor))

multiplot(
    ggplot(df) + geom_point(aes(x=x, y=y, color=donor)) +
        xlab("t-SNE 1") + ylab("t-SNE 2") + ggtitle("By donor"),
    ggplot(df) + geom_point(aes(x=x, y=y, color=batch)) +
        xlab("t-SNE 1") + ylab("t-SNE 2") + ggtitle("By batch"),
    cols=2)
```

## Merging the CEL-seq data

We directly merge together the two CEL-seq(2)-based data sets,
equivalent to our approach in the `r Biocpkg("simpleSingleCell", "batch.html", "previous workflow")`.
We note that each of these data sets also contains some donor-based structure,
but we will ignore it as it does not seem to interfere with the merge.

```{r}
mnn.umi <- fastMNN(pcs$gse81076, pcs$gse85241, k=20, pc.input=TRUE)
```

Figure \@ref(fig:umi-merge) demonstrates that these two data sets are successfully merged.
The quality of this merge is probably due to the fact that both data sets were generated by the same provider,
combined with the reduction in technical variability that is offered by UMI-based protocols.

```{r umi-merge, fig.cap="t-SNE plot of the merged CEL-seq data sets. Each point is a cell coloured by the batch of origin."}
set.seed(2000)
tsne.out <- Rtsne::Rtsne(mnn.umi$corrected, perplexity=30, pca=FALSE)
df <- data.frame(x=tsne.out$Y[,1], y=tsne.out$Y[,2], 
    batch=rep(c("GSE81076", "GSE85241"), 
        c(ncol(sce.gse81076), ncol(sce.gse85241))))

ggplot(df) + geom_point(aes(x=x, y=y, color=batch)) +
    xlab("t-SNE 1") + ylab("t-SNE 2") + ggtitle("By batch")
```

Finally, we merge the merged data sets across technologies by calling `fastMNN()` on the output of the per-technology merges.
This represents the final level of the hierarchical merge.

```{r}
mnn.final <- fastMNN(mnn.umi, mnn.smart, k=20, pc.input=TRUE)
```

Cells group together in distinct clusters that are consistent with the annotations from individual studies (Figure \@ref(fig:overall-merge)).
This suggests that the combined merge is satisfactory.

```{r overall-merge, fig.wide=TRUE, fig.asp=0.5, fig.cap="t-SNE plot of all merged pancreas data sets. Each point is a cell coloured by the batch of origin (left) or the annotated cell type (right)."}
set.seed(2000)
tsne.out <- Rtsne::Rtsne(mnn.final$corrected, perplexity=30, pca=FALSE)
df <- data.frame(x=tsne.out$Y[,1], y=tsne.out$Y[,2], 
    batch=rep(c("GSE81076", "GSE85241", "GSE86469", "E-MTAB-5061"), 
        c(ncol(sce.gse81076), ncol(sce.gse85241),
            ncol(sce.gse86469), ncol(sce.emtab))),
    type=c(rep("unknown", ncol(sce.gse81076)+ncol(sce.gse85241)), 
        sce.gse86469$CellType, sce.emtab$CellType),
    stringsAsFactors=FALSE)

# Restricting colors to certain approved cell types for visibility.
approved <- c(Acinar="#ffff00", Alpha="#ff0000", Beta="#c400ff", 
    Delta="#ff7800", Ductal="#00f5ff", Gamma="#0000ff",
    Other="#000000", unknown="grey80")
df$type[df$type=="Gamma/PP"] <- "Gamma"
df$type[!df$type %in% names(approved)] <- "Other"

multiplot(
    ggplot(df) + geom_point(aes(x=x, y=y, color=batch)) +
        xlab("t-SNE 1") + ylab("t-SNE 2") + ggtitle("By batch"),
    ggplot(df) + geom_point(aes(x=x, y=y, color=type)) +
        scale_color_manual(values=approved) +
        xlab("t-SNE 1") + ylab("t-SNE 2") + ggtitle("By cell type"),
    cols=2)
```

# Variance-based diagnostics 

As mentioned `r Biocpkg("simpleSingleCell", "batch.html#with-diagnostics", "previously")`, the proportion of lost variance within each batch serves as a useful diagnostic for whether biological structure was inadvertently removed during correction.
Here, the proportion lost at each merge step is low, indicating that the variation within each batch is mostly preserved.

```{r}
summary(metadata(fixed.emtab)$merge.info$lost.var)
metadata(mnn.smart)$merge.info$lost.var
metadata(mnn.umi)$merge.info$lost.var
metadata(mnn.final)$merge.info$lost.var
```

In a hierarchical merge, an additional metric is the proportion of variation lost during re-orthogonalization.
For example, consider `mnn.final` that results from merging `mnn.smart` and `mnn.umi`.
The orthogonalization step that was performed in `mnn.umi` is applied to the corrected values in `mnn.smart` and vice versa,
to ensure both inputs are processed in a consistent manner before attempting to merge them.
(In particular, it aims to avoid differences in variance that could interfere with a successful merge,
and/or compromise the utility of the proportion of variance as a diagnostic.)

Each re-orthogonalization step discards a proportion of variance that is recorded in the metadata of the `fastMNN()` output.
The intepretation of these propotions are the same as those from `merge.info$lost.var`, i.e.,
large proportions suggest that biological structure was in appropriate removed.
We can inspect the proportions at each of the "higher level" merge steps to ensure that this is not the case.

```{r}
summary(metadata(mnn.smart)$pre.orthog$lost.var)
summary(metadata(mnn.final)$pre.orthog$lost.var)
```

# Obtaining corrected expression values

If `fastMNN()` was run on low-dimensional inputs, only the low-dimensional output will be reported.
Nonetheless, users can obtain per-gene corrected values by manually computing the cross-product using the PCA rotation vectors.
For example, the code below obtains corrected expression values for _GCG_ from our hierarchical merge.

```{r}
rotations <- metadata(pcs)$rotation
cor.exp <- tcrossprod(mnn.final$corrected,
    rotations["ENSG00000115263",,drop=FALSE])
summary(cor.exp)
```

Explicit calculation of all per-gene corrected values is probably ill-advised as this would involve the construction of a dense matrix.
This may be prohibitively memory-consuming for large data sets that are otherwise representable as sparse matrices.
Rather, corrected values can be computed for specific genes as they are needed, e.g., using the `LowRankMatrix` class.

```{r}
lrm <- LowRankMatrix(rotations, mnn.final$corrected)
lrm
```

Users are referred to the `r Biocpkg("simpleSingleCell", "batch.html", "previous workflow")` for some caveats on the direct use of the corrected expression values.

# References
